contrib/pkg/utils/aws/aws.go

```diff
@@ -1,16 +1,227 @@
 package aws
 
 import (
+	"errors"
+	"fmt"
 	"os"
 	"path/filepath"
+	"strings"
 
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
 	"github.com/openshift/hive/contrib/pkg/utils"
+	"github.com/openshift/hive/pkg/awsclient"
 	"github.com/openshift/hive/pkg/constants"
+
 	log "github.com/sirupsen/logrus"
 	ini "gopkg.in/ini.v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+const (
+	// PrivateLinkHubAcctCredLabel is added to the AWS PrivateLink Hub account credentials Secret
+	// which is created by the "hiveutil awsprivatelink enable ..." command
+	// and referred to by HiveConfig.spec.awsPrivateLink.credentialsSecretRef.
+	PrivateLinkHubAcctCredLabel = "hive.openshift.io/awsprivatelink-hub-acct-credentials"
+)
+
+// AssociatedVpcWithClient encapsulates an associated VPC (for PrivateLink) and a set of AWS clients for it.
+type AssociatedVpcWithClient struct {
+	AWSClients    awsclient.Client
+	AssociatedVpc *hivev1.AWSPrivateLinkVPC
+}
+
+func FindVpcInInventory(vpcId string, inventory []hivev1.AWSPrivateLinkInventory) (endpointVpcIdx int, found bool) {
+	for i, endpointVpc := range inventory {
+		if vpcId == endpointVpc.AWSPrivateLinkVPC.VPCID {
+			endpointVpcIdx = i
+			found = true
+			break
+		}
+	}
+
+	return endpointVpcIdx, found
+}
+
+// GetDefaultSGOfVpc gets the default SG of a VPC.
+func GetDefaultSGOfVpc(awsClients awsclient.Client, vpcId *string) (string, error) {
+	describeSecurityGroupsOutput, err := awsClients.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{vpcId},
+			},
+			{
+				Name:   aws.String("group-name"),
+				Values: []*string{aws.String("default")},
+			},
+		},
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(describeSecurityGroupsOutput.SecurityGroups) == 0 {
+		return "", fmt.Errorf("default SG not found for VPC %v", vpcId)
+	}
+
+	return *describeSecurityGroupsOutput.SecurityGroups[0].GroupId, nil
+}
+
+// GetAssociatedVpcsWithClient extracts the associated VPCs from HiveConfig and returns a slice of AssociatedVpcWithClient.
+func GetAssociatedVpcsWithClient(hiveConfig *hivev1.HiveConfig) (associatedVpcsWithClient []*AssociatedVpcWithClient, err error) {
+	if hiveConfig.Spec.AWSPrivateLink == nil {
+		err = errors.New("AWS PrivateLink is not enabled in HiveConfig. ")
+		return
+	}
+
+	var awsClients awsclient.Client
+	for _, associatedVpc := range hiveConfig.Spec.AWSPrivateLink.AssociatedVPCs {
+		awsClients, err = awsclient.NewClientFromSecret(nil, associatedVpc.AWSPrivateLinkVPC.Region)
```

Comment:
Is the common case for each VPC to be in a different region? If it's possible to hit this line with the same region multiple times, I think we would be better off reusing the existing client for that region. Maybe use a local map to cache clients by region?

Reply:
I guess it is pretty common for an associate VPC or an endpoint VPC to be in the same region as another associated VPC or endpoint VPC. So yeah, it would be great if we could re-use the same client for each region.

Reply:
Now implemented.